src/Storages/MergeTree/MergeTreeBackgroundExecutor.h

```diff
@@ -15,10 +16,123 @@
 #include <Common/ThreadPool.h>
 #include <Common/Stopwatch.h>
 #include <Storages/MergeTree/IExecutableTask.h>
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
 
+struct TaskRuntimeData;
+using TaskRuntimeDataPtr = std::shared_ptr<TaskRuntimeData>;
 
-namespace DB
+/**
+ * Has RAII class to determine how many tasks are waiting for the execution and executing at the moment.
+ * Also has some flags and primitives to wait for current task to be executed.
+ */
+struct TaskRuntimeData
+{
+    TaskRuntimeData(ExecutableTaskPtr && task_, CurrentMetrics::Metric metric_)
+        : task(std::move(task_))
+        , increment(std::move(metric_))
+    {}
+
+    ExecutableTaskPtr task;
+    CurrentMetrics::Increment increment;
+    std::atomic_bool is_currently_deleting{false};
+    /// Actually autoreset=false is needed only for unit test
+    /// where multiple threads could remove tasks corresponding to the same storage
+    /// This scenario in not possible in reality.
+    Poco::Event is_done{/*autoreset=*/false};
+
+    /// By default priority queue will have max element at top
+    static bool comparePtrByPriority(const TaskRuntimeDataPtr & lhs, const TaskRuntimeDataPtr & rhs)
+    {
+        return lhs->task->getPriority() > rhs->task->getPriority();
```

Comment:
A lot of virtual calls, but maybe ok.
Alternative is to add `size_t priority` field and update it every time task is pushed to queue.